### PR TITLE
Tested Point3D

### PR DESCRIPTION
**Pull request by mcdaniel67**
(in geometry/plane.py)
Changed the NotImplementedError for collinear points to a ValueError.

(in geometry/point3d.py)
Removed elif statement from Point3D constructor which was satisfied by the initial if statement.

Changed the NotImplementedError to TypeError in the are_coplanar function, as suggested by comments.

Fixed a bug in the scale function, which caused the function to return without scaling.

Fixed a bug in the transformation function. It should take a 4x4 transformation matrix, and apply it to the 3d point with a homogeneous coordinate added. Previously, the function caused the z coordinate to be zeroed.

Added tests to bring the line coverage of point3d up to 99%

### CODE REVIEW

**Comment by debugger22** [comment]
@akshayah3 Could you please review this?

**Comment by akshayah3** [comment]
@mcdaniel67 Thanks for this! The changes look good to me.

**Comment by debugger22** [comment]
+1

**Comment by mcdaniel67** [comment]
No problem! I was wondering, what is the status of this request?

**Comment by moorepants** [comment]
We just have a reviewing queue. This looks good but I have one question.

File: sympy/geometry/point3d.py [hunk 1]
```diff
@@ -508,7 +506,7 @@ def translate(self, x=0, y=0, z=0):
 
     def transform(self, matrix):
         """Return the point after applying the transformation described
-        by the 3x3 Matrix, ``matrix``.
+        by the 4x4 Matrix, ``matrix``.
```
Comment (moorepants) [inline sympy/geometry/point3d.py]: This seems to be changing the functionality of a public method. Previously just the rotation matrix was returned, I assume? And now a transformation matrix is?

If this a necessary change it needs to be deprecated.

https://github.com/sympy/sympy/wiki/Deprecating-policy

Another option would be to have a new method that returns a transformed point.

Can you explain this change.

Comment (moorepants) [inline sympy/geometry/point3d.py]: I just read this:

> Fixed a bug in the transformation function. It should take a 4x4 transformation matrix, and apply it to the 3d point with a homogeneous coordinate added. Previously, the function caused the z coordinate to be zeroed.

So it sounds like this was a bug and didn't work in the first place. Can you confirm?

File: sympy/geometry/point3d.py [hunk 2]
```diff
@@ -77,8 +77,7 @@ def __new__(cls, *args, **kwargs):
             if not eval:
                 return args[0]
             args = args[0].args
-        elif isinstance(args[0], Point):
-            args = args[0].args
+        # elif was satisfied by the initial if statement and thus redundant
```
Comment (moorepants) [inline sympy/geometry/point3d.py]: No need to leave this comment in the code.

**Comment by mcdaniel67** [comment]
Yes, it was a bug that didn't work in the first place.

Previously, it was returning a value that was Point3D. One example of the bug would be if we transformed Point3D(1, 1, 1) by the identity matrix. Mathematically, it should return the same point, but the code actually returned Point3D(1, 1, 0). The zeroing of the z coordinate would happen regardless of the transformation matrix applied.

The existing code was applying a 3x3 matrix to a 3D point. This isn't how 3D transformations should be applied. Here is a link about 3D transformations: http://www.math.tau.ac.il/~dcor/Graphics/cg-slides/geom3d.pdf

The existing code did not even agree with the transformation function present in Point, which applies a 3x3 matrix to the 2D point with the homogeneous coordinate added.  (which is also wrong, as it applies the transpose of the matrix to the point. But that is an issue I've already worked out for another pull request)

**Comment by moorepants** [comment]
I'm not familiar with this package in SymPy so excuse my ignorance. I just want to make sure that the original `transform` method was supposed to rotate and translate using a 4 x 4 translation matrix or what is simply intended to rotate with a 3 x 3. There is a `translate` method so it seems to me that maybe they were setup to be independent operations. If so then maybe the translation and rotations should stay independent but the bug should still be fixed.

**Comment by mcdaniel67** [comment]
Ah, I see what you're saying, and I hadn't thought of it that way. The updated code definitely works for 4x4 rotations (just tested rotations about x, y, and z axes). Speed wise, it makes sense to keep the functions separate. (summing 3 values vs multiplying by a matrix)

Let me know if there is anything that you need me to do!

**Comment by moorepants** [comment]
I'm happy to merge this if you can confirm that `.transform()` was originally intended to do full transformations, i.e. rotations, translations, and scaling. If it was, then this can be merged after removing the unneeded comment I mentioned. But if `.transform()` was originally intended to simply rotate with a 3 x 3 matrix, then we may need to think of a different solution that respects SymPy's deprecation policy.

**Comment by mcdaniel67** [comment]
Would the best way for me to do that be to contact the person who added Point3D to SymPy? I can absolutely remove that comment, I wan't sure if I needed to supply a reasoning for deleting lines.

**Comment by jksuom** [comment]
It looks like Point3d was created by taking a copy of Point with its methods and adding one more coordinate in appropriate places. The ``transform`` in Point is given by a 3x3 matrix and includes a translation. It would seem illogical if this were the only method of Point that was intended to be treated differently in Point3d. (Even the See also part includes ``translate``.)

**Comment by moorepants** [comment]
Ok, that works. Thanks for clarifying. Remove that one comment and I'm +1 for merging.

**Comment by moorepants** [comment]
I see that your removed it. Thanks for the contribution.

**Pull request merged**